File: Advanced Methods of Programming/JavaClientServer - Spring/Persistence/src/main/java/Repo/RepoParticipant.java
package Repo;

import Model.Participant;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.ArrayList;
import java.util.List;
import java.util.Properties;

public class RepoParticipant implements IRepository<Integer, Participant> {
    private JdbcUtils dbUtils;

    private static final Logger logger= LogManager.getLogger();

    public RepoParticipant(Properties props){
        logger.info("Initializing RepoParticipant with properties: {} ",props);
        this.dbUtils=new JdbcUtils(props);
    }


    public void save(Participant participant) {
        logger.traceEntry("saving participant{}",participant);
        Connection con=dbUtils.getConnection();
        try(PreparedStatement preStmt=con.prepareStatement("insert into participant values (?,?,?)")){
            preStmt.setInt(1,participant.getId());
            preStmt.setString(2,participant.getNumeParticipant());
            preStmt.setInt(3,participant.getVarstaParticipant());
            preStmt.executeUpdate();
        }catch (SQLException ex){
            logger.error(ex);
            System.out.println("Error DB "+ex);
        }
        logger.traceExit();
    }

    @Override
    public int size() {
        logger.traceEntry();
        Connection con=dbUtils.getConnection();
        try(PreparedStatement preStmt=con.prepareStatement("select count(*) as SIZE from participant")) {
            try(ResultSet result = preStmt.executeQuery()) {
                if (result.next()) {
                    logger.traceExit(result.getInt("SIZE"));
                    return result.getInt("SIZE");
                }
            }
        }catch(SQLException ex){
            logger.error(ex);
            System.out.println("Error DB "+ex);
        }
        return 0;
    }

    @Override
    public Participant findOne(Integer idParticipant) {
        logger.traceEntry("finding participant with id {}",idParticipant);
        Connection con=dbUtils.getConnection();
        try(PreparedStatement preStmt=con.prepareStatement("select * from participant where idParticipant=?")) {
            preStmt.setInt(1,idParticipant);
            try(ResultSet result = preStmt.executeQuery()) {
                if (result.next()) {
                    Participant p= new Participant(result.getInt("idParticipant"),
                            result.getString("numeParticipant"), result.getInt("varstaParticipant"));
                    logger.traceExit(p);
                    return p;
                }
                else return null;
            }
        }catch(SQLException ex){
            logger.error(ex);
            System.out.println("Error DB "+ex);
        }
        logger.traceExit("No participant found with id{}",idParticipant);
        return null;
    }

    @Override
    public Iterable<Participant> findAll() {
        logger.traceEntry();
        List<Participant> lst = new ArrayList<>();
        Connection con=dbUtils.getConnection();
        try(PreparedStatement preStmt=con.prepareStatement("select * from participant")) {
            try(ResultSet result = preStmt.executeQuery()) {
                while (result.next()) {
                    lst.add(new Participant(result.getInt("idParticipant"), result.getString("numePArticipant"),
                            result.getInt("varstaParticipant")));
                }
            }
        }catch(SQLException ex){
            logger.error(ex);
            System.out.println("Error DB "+ex);
        }
        logger.traceExit(lst);
        return lst;
    }

    public Participant findByNameAndAge(String name, int age){
        logger.traceEntry("finding participant with name {} and age {}",name,age);
        Connection con=dbUtils.getConnection();
        try(PreparedStatement preStmt=con.prepareStatement("select * from participant where numeParticipant like ? and varstaParticipant=?")) {
            preStmt.setString(1,name);
            preStmt.setInt(2,age);
            try(ResultSet result = preStmt.executeQuery()) {
                if (result.next()) {
                    Participant participant= new Participant(result.getInt("idParticipant"),
                            result.getString("numeParticipant"), result.getInt("varstaParticipant"));
                    logger.traceExit(participant);
                    return participant;
                }
                else return null;
            }
        }catch(SQLException ex){
            logger.error(ex);
            System.out.println("Error DB "+ex);
        }
        logger.traceExit("No participant found with given name and age");
        return null;
    }

}

File: Advanced Methods of Programming/Client-ServerCSharp/Server/Repository/RepoParticipant.cs
using System;
using System.Collections.Generic;
using System.Data;
using SharedDLL.Domain;

namespace Server.Repo
{
    public class RepoParticipant : IRepository<int, Participant>
    {
        //private static readonly ILog log = LogManager.GetLogger("SortingTaskDbRepository");

        IDictionary<String, string> props;
        public RepoParticipant(IDictionary<String,string> props)
        {
            //log.Info("Creating RepoParticipants ");
            this.props = props;
        }
        public IEnumerable<Participant> FindAll()
        {
            IDbConnection con = DBUtils.getConnection(props);
            IList<Participant> participanti = new List<Participant>();
            using (var comm = con.CreateCommand())
            {
                comm.CommandText = "select idParticipant,numeParticipant,varstaParticipant from participant";

                using (var dataR = comm.ExecuteReader())
                {
                    while (dataR.Read())
                    {
                        int idParticipant = dataR.GetInt32(0);
                        string numeParticipant = dataR.GetString(1);
                        int varstaParticipant = dataR.GetInt32(2);
                        Participant participant = new Participant(idParticipant, numeParticipant, varstaParticipant);
                        participanti.Add(participant);
                    }
                }
            }

            return participanti;
        }

        public Participant FindOne(int id)
        {
            IDbConnection con = DBUtils.getConnection(props);
            IList<Participant> participanti = new List<Participant>();
            using (var comm = con.CreateCommand())
            {   
                comm.CommandText = "select idParticipant,numeParticipant,varstaParticipant " +
                    "from participant where idParticipant=@idParticipant";
                IDbDataParameter paramId = comm.CreateParameter();
                paramId.ParameterName = "@idParticipant";
                paramId.Value = id;
                comm.Parameters.Add(paramId);

                using (var dataR = comm.ExecuteReader())
                {
                    while (dataR.Read())
                    {
                        int idParticipant = dataR.GetInt32(0);
                        string numeParticipant = dataR.GetString(1);
                        int varstaParticipant = dataR.GetInt32(2);
                        Participant participant = new Participant(idParticipant, numeParticipant, varstaParticipant);
                        //log.InfoFormat("Exiting findOne with value {0}", task);
                        return participant;
                    }
                }
            }
            //log.InfoFormat("Exiting findOne with value {0}", null);
            return null;
        }

        public void Save(Participant participant)
        {
            var con = DBUtils.getConnection(props);

            using (var comm = con.CreateCommand())
            {
                comm.CommandText = "insert into participant  values (@id, @nume, @varsta)";
                var paramId = comm.CreateParameter();
                paramId.ParameterName = "@id";
                paramId.Value = participant.GetId();
                comm.Parameters.Add(paramId);

                var paramName = comm.CreateParameter();
                paramName.ParameterName = "@nume";
                paramName.Value = participant.GetNumeParticipant();
                comm.Parameters.Add(paramName);

                var paramVarsta = comm.CreateParameter();
                paramVarsta.ParameterName = "@varsta";
                paramVarsta.Value = participant.GetVarstaParticipant();
                comm.Parameters.Add(paramVarsta);

                var result = comm.ExecuteNonQuery();
                if (result == 0)
                    throw new RepositoryException("Nu s-a adaugat niciun angajat !");
            }
        }

        public int Size()
        {
            IDbConnection con = DBUtils.getConnection(props);
            using (var comm = con.CreateCommand())
            {
                comm.CommandText = "select count(*) from participant";
                using (var dataR = comm.ExecuteReader())
                {
                    if (dataR.Read())
                    {
                        //log.InfoFormat("Exiting findOne with value {0}", task);
                        return dataR.GetInt32(0);
                    }
                }
            }
            //log.InfoFormat("Exiting findOne with value {0}", null);
            return 0;
        }

        public int MaxId()
        {
            IDbConnection con = DBUtils.getConnection(props);
            using (var comm = con.CreateCommand())
            {
                comm.CommandText = "select max(idParticipant) from participant";
                using (var dataR = comm.ExecuteReader())
                {
                    if (dataR.Read())
                    {
                        //log.InfoFormat("Exiting findOne with value {0}", task);
                        return dataR.GetInt32(0);
                    }
                }
            }
            //log.InfoFormat("Exiting findOne with value {0}", null);
            return 0;
        }
    }
}
